### PR TITLE
Demo loop updates

### PR DESCRIPTION
**Pull request by amandalund**
This starts filling in some pieces of the `demo-loop` app, including some of the kernels (pre-/post-/along-step, model launches, secondary and interaction post-processing). It's still a work in progress and obviously untested since we aren't able to run the app yet, but it should be ok as a jumping-off point for our upcoming hackathon. I've also added a small fix to make sure the prime energy index is found correctly and updated the `SimInterface` to use `StateCollection`.

### CODE REVIEW

File: app/demo-loop/KernelUtils.i.hh [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                                   const PhysicsTrackView& phys)
+{
+    // Actual distance, limited by along-step length or geometry
+    real_type step = phys.step_length();
+    if (step > 0)
+    {
+        // Propagate up to the step length or next boundary
+        LinearPropagator propagate(&geo);
+        auto             geo_step = propagate(step);
+        step                      = geo_step.distance;
```
Comment (sethrj) [inline app/demo-loop/KernelUtils.i.hh:62]: Here we'll need to add code to check whether the volume/material have changed.

**Review by sethrj — COMMENTED** [review]
Thanks @amandalund , looks very good. Since this PR is about incremental improvements and not completeness, just to be clear my comments aren't asking for new capability but just little reminders :)

One other question was how/why the ROOT file was changed as part of this PR? Is that a rebase issue?

File: app/CMakeLists.txt [hunk 2]
```diff
@@ -186,7 +186,8 @@ endif()
 #-----------------------------------------------------------------------------#
 # DEMO: full physics loop
 #-----------------------------------------------------------------------------#
-if(CELERITAS_BUILD_DEMOS AND CELERITAS_USE_VecGeom)
+
+if(CELERITAS_BUILD_DEMOS AND CELERITAS_USE_VecGeom AND CELERITAS_USE_ROOT)
```
[the remark this answers is not in the document]
Comment (amandalund) [reply app/CMakeLists.txt:190]: ahh you're right, thanks!

File: app/demo-loop/KernelUtils.i.hh (hunk 1 above)
Comment (amandalund) [reply app/demo-loop/KernelUtils.i.hh:62]: Added a reminder :)

**Comment by amandalund** [comment]
> One other question was how/why the ROOT file was changed as part of this PR? Is that a rebase issue?

Oops thanks, I'll switch that back (I made a temporary ROOT input with only the exported data for the processes/particles in the problem  when I was debugging things).

**Review by sethrj — APPROVED** [review]
Looks good, thanks!

**Pull request merged**